sdk/emm/ema.c

```diff
@@ -0,0 +1,1177 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <string.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include "ema.h"
+#include "bit_array.h"
+#include "sgx_mm.h"
+#include "sgx_mm_primitives.h"
+#include "sgx_mm_rt_abstraction.h"
+/* State flags */
+#define SGX_EMA_STATE_PENDING   0x8UL
+#define SGX_EMA_STATE_MODIFIED  0x10UL
+#define SGX_EMA_STATE_PR        0x20UL
+#define ROUND_TO(x, align)        (((x) + (align-1)) & ~(align-1))
+#define TRIM_TO(x, align)        ((x) & ~(align-1))
+#define MIN(x, y) (((x)>(y))?(y):(x))
+#define MAX(x, y) (((x)>(y))?(x):(y))
+#define UNUSED(x) ((void)(x))
+struct ema_t_ {
+    size_t              start_addr;     // starting address, should be on a page boundary
+    size_t              size;           // bytes
+    uint32_t            alloc_flags;    // EMA_RESERVED, EMA_COMMIT_NOW, EMA_COMMIT_ON_DEMAND,
+                                        // OR'ed with EMA_SYSTEM, EMA_GROWSDOWN, ENA_GROWSUP
+    uint64_t            si_flags;       // one of EMA_PROT_NONE, READ, READ_WRITE, READ_EXEC, READ_WRITE_EXEC
+                                        // Or'd with one of EMA_PAGE_TYPE_REG, EMA_PAGE_TYPE_TCS, EMA_PAGE_TYPE_TRIM
+    bit_array *         eaccept_map;    // bitmap for EACCEPT status, bit 0 in eaccept_map[0] for the page at start address
+                                        // bit i in eaccept_map[j] for page at start_address+(i+j<<3)<<12
+    sgx_mm_mutex *      lock;           // lock to prevent concurrent modification, could be sgx_thread_mutex_t/rwlock_t
+    int                 transition;     // state to indicate whether a transition in progress, e.g page type/permission changes.
+    sgx_enclave_fault_handler_t
+                        handler;              // custom PF handler  (for EACCEPTCOPY use)
+    void*               private;        // private data for handler
+    ema_t*              next;           // next in doubly linked list
+    ema_t*              prev;           // prev in doubly linked list
+};
+
+struct ema_root_ {
+    ema_t *guard;
+};
+
+ema_t dummy_user_ema = {.next = &dummy_user_ema,
+                        .prev = &dummy_user_ema};
+ema_root_t g_user_ema_root = {.guard = &dummy_user_ema};
+
+ema_t dummy_rts_ema = {.next = &dummy_rts_ema,
+                       .prev = &dummy_rts_ema};
+ema_root_t g_rts_ema_root = {.guard = &dummy_rts_ema};
+
+#ifdef TEST
+static void dump_ema_node(ema_t *node, size_t index)
+{
+    printf("------ node #%lu ------\n", index);
+    printf("start:\t0x%lX\n", node->start_addr);
+    printf("size:\t0x%lX\n", node->size);
+}
+
+void dump_ema_root(ema_root_t *root)
+{
+    ema_t *node = root->guard->next;
+    size_t index = 0;
+
+    while (node != root->guard) {
+        dump_ema_node(node, index++);
+        node = node->next;
+    }
+}
+
+#endif
+void destroy_ema_root(ema_root_t *root)
```

Comment:
it should be invoked right before sgx_detroy_enclave?

Reply:
When enclave is destroyed/gone for any reason (fd to /dev/sgx_enclave is closed, host app crash/exit, etc.), the kernel will release whole range and EPC allocation. So there is really no need to cleanup memory inside enclave. Although free-up ema  may also be OK, we need be careful about the order relative to other cleanup (e.g., TCS), ema should be the very last to go.

Reply:
ok , i understand. thanks, haitao.